qmail-smtpd.c

```diff
@@ -29,13 +29,7 @@
 unsigned int databytes = 0;
 int timeout = 1200;
 
-ssize_t safewrite(int fd, const void *buf, size_t len)
-{
-  ssize_t r;
-  r = timeoutwrite(timeout,fd,buf,len);
-  if (r == 0 || r == -1) _exit(1);
-  return r;
-}
+GEN_SAFE_TIMEOUTWRITE(safewrite,fd,timeout,_exit(1))
```

Comment:
Before this commit, `make qmail-smtpd && ./qmail-smtpd` gives `421 unable to read controls (#4.3.0)`. After this commit, it exits 1 silently. I don't understand yet why this change has this effect, but it does.

Reply:
Ah, now I get it. Drop the `fd` argument and have the macro call `timeoutwrite(tout,fd,buf,len)`, not `timeoutwrite(tout,writefd,buf,len)`.

Reply:
The macro actually has to be that way: there are several instances where the fd passed to the function is ignored and a local variable is used, e.g. in qmail-remote.c. The bug is that I swapped the arguments to the macro, it must be ```GEN_SAFE_TIMEOUTWRITE(safewrite,timeout,fd,_exit(1))```

Reply:
Yep, I see it now. I may never get my `cpp` license.